panic: runtime error: invalid memory address or nil pointer dereference

初次使用，添加一个config成功后，再次添加开始报错，kubecm ls也报，如下：

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x1ea0aa5]

goroutine 1 [running]:
github.com/sunny0826/kubecm/cmd.Formatable(0xc0002d2c50, 0xc000127d58)
	/github/workspace/cmd/utils.go:97 +0x285
github.com/sunny0826/kubecm/cmd.(*ListCommand).runList(0xc0002bf570, 0xc000263b80, 0x2b28a18, 0x0, 0x0, 0x0, 0x0)
	/github/workspace/cmd/list.go:30 +0x26
github.com/sunny0826/kubecm/cmd.(*ListCommand).Init.func1(0xc000263b80, 0x2b28a18, 0x0, 0x0, 0x0, 0x0)
	/github/workspace/cmd/list.go:22 +0x52
github.com/spf13/cobra.(*Command).execute(0xc000263b80, 0x2b28a18, 0x0, 0x0, 0xc000263b80, 0x2b28a18)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:842 +0x47c
github.com/spf13/cobra.(*Command).ExecuteC(0xc0002622c0, 0x10449b7, 0x2a847a0, 0xc000058778)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
main.main()
	/github/workspace/main.go:27 +0x36

--- a/cmd/add.go
+++ b/cmd/add.go
@@ -67,6 +67,7 @@ func formatNewConfig(file string) (*clientcmdapi.Config, string, error) {
 		return nil, "", errors.New("Only support add 1 context. You can use `merge` cmd")
 	}
 	name, err := formatAndCheckName(file)
+	config = CheckValidContext(config)
 	if err != nil {
 		return nil, "", err
 	}

--- a/cmd/clear.go
+++ b/cmd/clear.go
@@ -2,10 +2,11 @@ package cmd
 
 import (
 	"fmt"
-	"github.com/spf13/cobra"
-	"k8s.io/client-go/tools/clientcmd"
 	"os"
 	"reflect"
+
+	"github.com/spf13/cobra"
+	"k8s.io/client-go/tools/clientcmd"
 )
 
 // ClearCommand clean command struct
@@ -54,7 +55,7 @@ func clearContext(file string) (bool, error) {
 	}
 	outConfig := config.DeepCopy()
 	outConfig = CheckValidContext(outConfig)
-	if reflect.DeepEqual(config,outConfig) {
+	if reflect.DeepEqual(config, outConfig) {
 		return true, nil
 	}
 	err = UpdateConfigFile(file, outConfig)

--- a/cmd/clear.go
+++ b/cmd/clear.go
@@ -1,0 +1,74 @@
+package cmd
+
+import (
+	"fmt"
+	"github.com/spf13/cobra"
+	"k8s.io/client-go/tools/clientcmd"
+	"os"
+	"reflect"
+)
+
+// ClearCommand clean command struct
+type ClearCommand struct {
+	BaseCommand
+}
+
+// Init AliasCommand
+func (cl *ClearCommand) Init() {
+	cl.command = &cobra.Command{
+		Use:   "clear",
+		Short: "Clear lapsed context, cluster and user",
+		Long:  "Clear lapsed context, cluster and user",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return cl.runClear(cmd, args)
+		},
+		Example: clearExample(),
+	}
+	cl.command.DisableFlagsInUseLine = true
+}
+
+func (cl *ClearCommand) runClear(cmd *cobra.Command, args []string) error {
+	if len(args) == 0 {
+		if ok, err := clearContext(cfgFile); ok {
+			printString(os.Stdout, fmt.Sprintf("There is nothing to clean in 「%s」\n", cfgFile))
+			return nil
+		} else if err != nil {
+			return err
+		}
+	} else {
+		for _, file := range args {
+			if ok, err := clearContext(file); ok {
+				printString(os.Stdout, fmt.Sprintf("There is nothing to clean in 「%s」\n", file))
+			} else if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func clearContext(file string) (bool, error) {
+	config, err := clientcmd.LoadFromFile(file)
+	if err != nil {
+		return false, err
+	}
+	outConfig := config.DeepCopy()
+	outConfig = CheckValidContext(outConfig)
+	if reflect.DeepEqual(config,outConfig) {
+		return true, nil
+	}
+	err = UpdateConfigFile(file, outConfig)
+	if err != nil {
+		return false, err
+	}
+	return false, nil
+}
+
+func clearExample() string {
+	return fmt.Sprintf(`
+# Clear lapsed context, cluster and user (default is %s)
+kubecm clear
+# Customised clear lapsed files
+kubecm clear config.yaml test.yaml
+`, cfgFile)
+}

--- a/cmd/clear_test.go
+++ b/cmd/clear_test.go
@@ -2,9 +2,10 @@ package cmd
 
 import (
 	"io/ioutil"
-	"k8s.io/client-go/tools/clientcmd"
 	"os"
 	"testing"
+
+	"k8s.io/client-go/tools/clientcmd"
 )
 
 func Test_clearContext(t *testing.T) {

--- a/cmd/clear_test.go
+++ b/cmd/clear_test.go
@@ -1,0 +1,43 @@
+package cmd
+
+import (
+	"io/ioutil"
+	"k8s.io/client-go/tools/clientcmd"
+	"os"
+	"testing"
+)
+
+func Test_clearContext(t *testing.T) {
+	trueFile, _ := ioutil.TempFile("", "")
+	falseFile, _ := ioutil.TempFile("", "")
+	defer os.Remove(trueFile.Name())
+	defer os.Remove(falseFile.Name())
+	_ = clientcmd.WriteToFile(appendMergeConfig, trueFile.Name())
+	_ = clientcmd.WriteToFile(wrongRootConfig, falseFile.Name())
+
+	type args struct {
+		file string
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    bool
+		wantErr bool
+	}{
+		// TODO: Add test cases.
+		{"true", args{file: trueFile.Name()}, true, false},
+		{"false", args{file: falseFile.Name()}, false, false},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := clearContext(tt.args.file)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("clearContext() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if got != tt.want {
+				t.Errorf("clearContext() got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}

--- a/cmd/cmd.go
+++ b/cmd/cmd.go
@@ -17,6 +17,7 @@ func NewBaseCommand() *BaseCommand {
 		&NamespaceCommand{},  // namespace command
 		&ListCommand{},       // list command
 		&AliasCommand{},      // alias command
+		&ClearCommand{},      // clear command
 	)
 
 	return baseCmd

--- a/cmd/list.go
+++ b/cmd/list.go
@@ -1,9 +1,13 @@
 package cmd
 
 import (
-	"log"
+	"errors"
+	"fmt"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"strings"
 
 	"github.com/spf13/cobra"
+	"k8s.io/client-go/tools/clientcmd"
 )
 
 // ListCommand list cmd struct
@@ -27,16 +31,47 @@ func (lc *ListCommand) Init() {
 }
 
 func (lc *ListCommand) runList(command *cobra.Command, args []string) error {
-	err := Formatable()
+	config, err := clientcmd.LoadFromFile(cfgFile)
+	if err != nil {
+		return err
+	}
+	config = CheckValidContext(config)
+	outConfig, err := filterArgs(args, config)
+	if err != nil {
+		return err
+	}
+	err = PrintTable(outConfig)
 	if err != nil {
 		return err
 	}
 	err = ClusterStatus()
 	if err != nil {
-		log.Fatalf("Cluster check failure!\n%v", err)
-		return err
+		return errors.New(fmt.Sprintf("Cluster check failure!\n%v", err))
 	}
 	return nil
+}
+
+func filterArgs(args []string, config *clientcmdapi.Config) (*clientcmdapi.Config, error) {
+	if len(args) == 0 {
+		return config, nil
+	}
+	contextList := make(map[string]string)
+	for key := range config.Contexts {
+		for _, search := range args {
+			if strings.Contains(key, search) {
+				contextList[key] = search
+			}
+		}
+	}
+	for key := range config.Contexts {
+		if _, ok := contextList[key]; !ok {
+			delete(config.Contexts, key)
+		}
+	}
+	if len(config.Contexts) == 0 {
+		return nil, errors.New(fmt.Sprintf("There is no matching context for %v\n", args))
+	}
+	return config, nil
 }
 
 func listExample() string {
@@ -45,5 +80,7 @@ func listExample() string {
 kubecm ls
 # Aliases
 kubecm l
+# Filter out keywords(Multi-keyword support)
+kubecm ls kind k3s
 `
 }

--- a/cmd/list.go
+++ b/cmd/list.go
@@ -69,7 +69,7 @@ func filterArgs(args []string, config *clientcmdapi.Config) (*clientcmdapi.Confi
 		}
 	}
 	if len(config.Contexts) == 0 {
-		return nil, fmt.Errorf("There is no matching context for %v\n", args)
+		return nil, fmt.Errorf("there is no matching context for %v", args)
 	}
 	return config, nil
 }

--- a/cmd/list.go
+++ b/cmd/list.go
@@ -1,10 +1,10 @@
 package cmd
 
 import (
-	"errors"
 	"fmt"
-	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 	"strings"
+
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 
 	"github.com/spf13/cobra"
 	"k8s.io/client-go/tools/clientcmd"
@@ -46,7 +46,7 @@ func (lc *ListCommand) runList(command *cobra.Command, args []string) error {
 	}
 	err = ClusterStatus()
 	if err != nil {
-		return errors.New(fmt.Sprintf("Cluster check failure!\n%v", err))
+		return fmt.Errorf("Cluster check failure!\n%v", err)
 	}
 	return nil
 }
@@ -69,7 +69,7 @@ func filterArgs(args []string, config *clientcmdapi.Config) (*clientcmdapi.Confi
 		}
 	}
 	if len(config.Contexts) == 0 {
-		return nil, errors.New(fmt.Sprintf("There is no matching context for %v\n", args))
+		return nil, fmt.Errorf("There is no matching context for %v\n", args)
 	}
 	return config, nil
 }

--- a/cmd/list_test.go
+++ b/cmd/list_test.go
@@ -1,0 +1,71 @@
+package cmd
+
+import (
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"reflect"
+	"testing"
+)
+
+var (
+	noRootMergeConfig = clientcmdapi.Config{
+		AuthInfos: map[string]*clientcmdapi.AuthInfo{
+			"black-user": {Token: "black-token"},
+			"red-user":   {Token: "red-token"},
+		},
+		Clusters: map[string]*clientcmdapi.Cluster{
+			"pig-cluster": {Server: "http://pig.org:8080"},
+			"cow-cluster": {Server: "http://cow.org:8080"},
+		},
+		Contexts: map[string]*clientcmdapi.Context{
+			"federal-context": {AuthInfo: "red-user", Cluster: "cow-cluster", Namespace: "hammer-ns"},
+		},
+	}
+	noFederalMergeConfig = clientcmdapi.Config{
+		AuthInfos: map[string]*clientcmdapi.AuthInfo{
+			"black-user": {Token: "black-token"},
+			"red-user":   {Token: "red-token"},
+		},
+		Clusters: map[string]*clientcmdapi.Cluster{
+			"pig-cluster": {Server: "http://pig.org:8080"},
+			"cow-cluster": {Server: "http://cow.org:8080"},
+		},
+		Contexts: map[string]*clientcmdapi.Context{
+			"root-context": {AuthInfo: "black-user", Cluster: "pig-cluster", Namespace: "saw-ns"},
+		},
+	}
+)
+
+func NewAllConfig() *clientcmdapi.Config {
+	return appendMergeConfig.DeepCopy()
+}
+
+func Test_filterArgs(t *testing.T) {
+
+	type args struct {
+		args   []string
+		config *clientcmdapi.Config
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    *clientcmdapi.Config
+		wantErr bool
+	}{
+		// TODO: Add test cases.
+		{"noFederal", args{[]string{"root"}, NewAllConfig()}, &noFederalMergeConfig, false},
+		{"noRoot", args{[]string{"federal"}, NewAllConfig()}, &noRootMergeConfig, false},
+		{"all-context", args{[]string{"root", "federal"}, NewAllConfig()}, &appendMergeConfig, false},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := filterArgs(tt.args.args, tt.args.config)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("filterArgs() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("filterArgs() got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}

--- a/cmd/list_test.go
+++ b/cmd/list_test.go
@@ -1,9 +1,10 @@
 package cmd
 
 import (
-	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 	"reflect"
 	"testing"
+
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 )
 
 var (

--- a/cmd/root.go
+++ b/cmd/root.go
@@ -17,9 +17,15 @@ limitations under the License.
 */
 
 import (
+	"bytes"
 	"flag"
+	"fmt"
 	"os"
+	"os/exec"
+	"os/user"
 	"path/filepath"
+	"runtime"
+	"strings"
 
 	"github.com/spf13/cobra"
 )
@@ -73,10 +79,49 @@ func (cli *Cli) Run() error {
 	}
 	return cli.rootCmd.Execute()
 }
-
 func homeDir() string {
-	if h := os.Getenv("HOME"); h != "" {
-		return h
+	u, err := user.Current()
+	if nil == err {
+		return u.HomeDir
 	}
-	return os.Getenv("USERPROFILE") // windows
+	// cross compile support
+	if "windows" == runtime.GOOS {
+		return homeWindows()
+	}
+	// Unix-like system, so just assume Unix
+	return homeUnix()
+}
+
+func homeUnix() string {
+	if home := os.Getenv("HOME"); home != "" {
+		return home
+	}
+	var stdout bytes.Buffer
+	cmd := exec.Command("sh", "-c", "eval echo ~$USER")
+	cmd.Stdout = &stdout
+	if err := cmd.Run(); err != nil {
+		return ""
+	}
+	result := strings.TrimSpace(stdout.String())
+	if result == "" {
+		fmt.Println("blank output when reading home directory")
+		os.Exit(0)
+	}
+
+	return result
+}
+
+func homeWindows() string {
+	drive := os.Getenv("HOMEDRIVE")
+	path := os.Getenv("HOMEPATH")
+	home := drive + path
+	if drive == "" || path == "" {
+		home = os.Getenv("USERPROFILE")
+	}
+	if home == "" {
+		fmt.Println("HOMEDRIVE, HOMEPATH, and USERPROFILE are blank")
+		os.Exit(0)
+	}
+
+	return home
 }

--- a/cmd/switch.go
+++ b/cmd/switch.go
@@ -53,10 +53,6 @@ func (sc *SwitchCommand) runSwitch(command *cobra.Command, args []string) error 
 		return err
 	}
 	sc.command.Printf("Switched to context 「%s」\n", config.CurrentContext)
-	//err = Formatable()
-	//if err != nil {
-	//	return err
-	//}
 	return nil
 }
 

--- a/cmd/utils.go
+++ b/cmd/utils.go
@@ -353,13 +353,13 @@ func CheckValidContext(config *clientcmdapi.Config) *clientcmdapi.Config {
 			printString(os.Stdout, "Check Config: ")
 			fmt.Printf("AuthInfo 「%s」 has no matching context 「%s」, skip\n", obj.AuthInfo, key)
 			delete(config.Contexts, key)
-			delete(config.Clusters,obj.Cluster)
+			delete(config.Clusters, obj.Cluster)
 		}
 		if _, ok := config.Clusters[obj.Cluster]; !ok {
 			printString(os.Stdout, "Check Config: ")
 			fmt.Printf("Cluster 「%s」 has no matching context 「%s」, skip\n", obj.Cluster, key)
 			delete(config.Contexts, key)
-			delete(config.AuthInfos,obj.AuthInfo)
+			delete(config.AuthInfos, obj.AuthInfo)
 		}
 	}
 	return config

--- a/cmd/utils.go
+++ b/cmd/utils.go
@@ -78,12 +78,8 @@ func HashSuf(config *clientcmdapi.Config) string {
 	return sum
 }
 
-// Formatable generate table
-func Formatable() error {
-	config, err := clientcmd.LoadFromFile(cfgFile)
-	if err != nil {
-		return err
-	}
+// PrintTable generate table
+func PrintTable(config *clientcmdapi.Config) error {
 	var table [][]string
 	for key, obj := range config.Contexts {
 		namespace := "default"
@@ -94,7 +90,8 @@ func Formatable() error {
 		if obj.Namespace != "" {
 			namespace = obj.Namespace
 		}
-		conTmp := []string{head, key, obj.Cluster, obj.AuthInfo, config.Clusters[obj.Cluster].Server, namespace}
+		server := config.Clusters[obj.Cluster].Server
+		conTmp := []string{head, key, obj.Cluster, obj.AuthInfo, server, namespace}
 		table = append(table, conTmp)
 	}
 
@@ -230,7 +227,7 @@ func WriteConfig(cover bool, file string, outConfig *clientcmdapi.Config) error 
 			return err
 		}
 		fmt.Printf("「%s」 write successful!\n", file)
-		err = Formatable()
+		err = PrintTable(outConfig)
 		if err != nil {
 			return err
 		}
@@ -241,6 +238,19 @@ func WriteConfig(cover bool, file string, outConfig *clientcmdapi.Config) error 
 		}
 		fmt.Println("generate ./config.yaml")
 	}
+	return nil
+}
+
+func UpdateConfigFile(file string, updateConfig *clientcmdapi.Config) error {
+	file, err := CheckAndTransformFilePath(file)
+	if err != nil {
+		return err
+	}
+	err = clientcmd.WriteToFile(*updateConfig, file)
+	if err != nil {
+		return err
+	}
+	printString(os.Stdout, "Update Config: "+file+"\n")
 	return nil
 }
 
@@ -334,4 +344,23 @@ func CheckAndTransformFilePath(path string) (string, error) {
 		return "", err
 	}
 	return path, nil
+}
+
+// CheckValidContext check and clean mismatched AuthInfo and Cluster
+func CheckValidContext(config *clientcmdapi.Config) *clientcmdapi.Config {
+	for key, obj := range config.Contexts {
+		if _, ok := config.AuthInfos[obj.AuthInfo]; !ok {
+			printString(os.Stdout, "Check Config: ")
+			fmt.Printf("AuthInfo 「%s」 has no matching context 「%s」, skip\n", obj.AuthInfo, key)
+			delete(config.Contexts, key)
+			delete(config.Clusters,obj.Cluster)
+		}
+		if _, ok := config.Clusters[obj.Cluster]; !ok {
+			printString(os.Stdout, "Check Config: ")
+			fmt.Printf("Cluster 「%s」 has no matching context 「%s」, skip\n", obj.Cluster, key)
+			delete(config.Contexts, key)
+			delete(config.AuthInfos,obj.AuthInfo)
+		}
+	}
+	return config
 }

--- a/cmd/utils_test.go
+++ b/cmd/utils_test.go
@@ -1,6 +1,7 @@
 package cmd
 
 import (
+	"os/user"
 	"reflect"
 	"testing"
 
@@ -31,6 +32,31 @@ var (
 		AuthInfos: map[string]*clientcmdapi.AuthInfo{
 			"black-user": {Token: "black-token"},
 			"red-user":   {Token: "red-token"},
+		},
+		Clusters: map[string]*clientcmdapi.Cluster{
+			"pig-cluster": {Server: "http://pig.org:8080"},
+			"cow-cluster": {Server: "http://cow.org:8080"},
+		},
+		Contexts: map[string]*clientcmdapi.Context{
+			"root-context":    {AuthInfo: "black-user", Cluster: "pig-cluster", Namespace: "saw-ns"},
+			"federal-context": {AuthInfo: "red-user", Cluster: "cow-cluster", Namespace: "hammer-ns"},
+		},
+	}
+	wrongRootConfig = clientcmdapi.Config{
+		AuthInfos: map[string]*clientcmdapi.AuthInfo{
+			"red-user": {Token: "red-token"},
+		},
+		Clusters: map[string]*clientcmdapi.Cluster{
+			"cow-cluster": {Server: "http://cow.org:8080"},
+		},
+		Contexts: map[string]*clientcmdapi.Context{
+			"root-context":    {AuthInfo: "black-user", Cluster: "pig-cluster", Namespace: "saw-ns"},
+			"federal-context": {AuthInfo: "red-user", Cluster: "cow-cluster", Namespace: "hammer-ns"},
+		},
+	}
+	wrongFederalConfig = clientcmdapi.Config{
+		AuthInfos: map[string]*clientcmdapi.AuthInfo{
+			"black-user": {Token: "black-token"},
 		},
 		Clusters: map[string]*clientcmdapi.Cluster{
 			"pig-cluster": {Server: "http://pig.org:8080"},
@@ -117,4 +143,91 @@ func testSetNilMapsToEmpties(curr reflect.Value) {
 
 	}
 
+}
+
+func TestExitOption(t *testing.T) {
+	gotNeedles := []Needle{
+		{"test1", "test2", "any", "*"},
+		{"test", "test2", "any", ""},
+	}
+	u, _ := user.Current()
+	wantNeedles := []Needle{
+		{"test1", "test2", "any", "*"},
+		{"test", "test2", "any", ""},
+		{"<Exit>", "exit the kubecm", u.Username, ""},
+	}
+	type args struct {
+		kubeItems []Needle
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    []Needle
+		wantErr bool
+	}{
+		// TODO: Add test cases.
+		{"test", args{gotNeedles}, wantNeedles, false},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := ExitOption(tt.args.kubeItems)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("ExitOption() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("ExitOption() got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestCheckAndTransformFilePath(t *testing.T) {
+	wantPath := homeDir() + "/.kube/config"
+	type args struct {
+		path string
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    string
+		wantErr bool
+	}{
+		// TODO: Add test cases.
+		{"test-~", args{path: "~/.kube/config"}, wantPath, false},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := CheckAndTransformFilePath(tt.args.path)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("CheckAndTransformFilePath() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if got != tt.want {
+				t.Errorf("CheckAndTransformFilePath() got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestCheckValidContext(t *testing.T) {
+	type args struct {
+		config *clientcmdapi.Config
+	}
+	tests := []struct {
+		name string
+		args args
+		want *clientcmdapi.Config
+	}{
+		// TODO: Add test cases.
+		{"clear-root", args{config: &wrongRootConfig}, &appendConfigAlfa},
+		{"clear-federal", args{config: &wrongFederalConfig}, &appendRootConfigConflictAlfa},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := CheckValidContext(tt.args.config); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("CheckValidContext() = %v, want %v", got, tt.want)
+			}
+		})
+	}
 }

--- a/cmd/utils_test.go
+++ b/cmd/utils_test.go
@@ -183,7 +183,7 @@ func TestExitOption(t *testing.T) {
 }
 
 func TestCheckAndTransformFilePath(t *testing.T) {
-	wantPath := homeDir() + "/.kube/config"
+	wantPath := homeDir()
 	type args struct {
 		path string
 	}
@@ -194,7 +194,7 @@ func TestCheckAndTransformFilePath(t *testing.T) {
 		wantErr bool
 	}{
 		// TODO: Add test cases.
-		{"test-~", args{path: "~/.kube/config"}, wantPath, false},
+		{"test-~", args{path: "~/"}, wantPath, false},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {

--- a/docs/en-us/README.md
+++ b/docs/en-us/README.md
@@ -23,6 +23,7 @@ Usage:
 Available Commands:
     add         Merge configuration file with $HOME/.kube/config
     alias       Generate alias for all contexts
+    clear       Clear lapsed context, cluster and user
     completion  Generates bash/zsh completion scripts
     delete      Delete the specified context from the kubeconfig
     help        Help about any command

--- a/docs/en-us/_sidebar.md
+++ b/docs/en-us/_sidebar.md
@@ -10,5 +10,6 @@
     * [kubecm namespace](/en-us/cli/kubecm_namespace.md)
     * [kubecm rename](/en-us/cli/kubecm_rename.md)
     * [kubecm switch](/en-us/cli/kubecm_switch.md)
+    * [kubecm clear](/en-us/cli/kubecm_clear.md)
     * [kubecm version](/en-us/cli/kubecm_version.md)
 * [Contribute](/en-us/contribute.md)

--- a/docs/en-us/cli/kubecm_clear.md
+++ b/docs/en-us/cli/kubecm_clear.md
@@ -1,0 +1,34 @@
+## kubecm clear
+
+Clear lapsed context, cluster and user
+
+### Synopsis
+
+Clear lapsed context, cluster and user
+
+```
+kubecm clear
+```
+
+### Examples
+
+```
+
+# Clear lapsed context, cluster and user (default is /Users/saybot/.kube/config)
+kubecm clear
+# Customised clear lapsed files
+kubecm clear config.yaml test.yaml
+
+```
+
+### Options
+
+```
+  -h, --help   help for clear
+```
+
+### Options inherited from parent commands
+
+```
+      --config string   path of kubeconfig (default "/Users/saybot/.kube/config")
+```

--- a/docs/en-us/cli/kubecm_ls.md
+++ b/docs/en-us/cli/kubecm_ls.md
@@ -18,17 +18,19 @@ kubecm ls
 kubecm ls
 # Aliases
 kubecm l
+# Filter out keywords(Multi-keyword support)
+kubecm ls kind k3s
 
 ```
 
 ### Options
 
 ```
-  -h, --help   help for ls
+  -h, --help            help for ls
 ```
 
 ### Options inherited from parent commands
 
 ```
-      --config string   path of kubeconfig (default "/Users/guoxudong/.kube/config")
+      --config string   path of kubeconfig (default "/Users/saybot/.kube/config")
 ```

--- a/docs/zh-cn/README.md
+++ b/docs/zh-cn/README.md
@@ -23,6 +23,7 @@ Usage:
 Available Commands:
     add         Merge configuration file with $HOME/.kube/config
     alias       Generate alias for all contexts
+    clear       Clear lapsed context, cluster and user
     completion  Generates bash/zsh completion scripts
     delete      Delete the specified context from the kubeconfig
     help        Help about any command

--- a/docs/zh-cn/_sidebar.md
+++ b/docs/zh-cn/_sidebar.md
@@ -10,5 +10,6 @@
     * [kubecm namespace](/zh-cn/cli/kubecm_namespace.md)
     * [kubecm rename](/zh-cn/cli/kubecm_rename.md)
     * [kubecm switch](/zh-cn/cli/kubecm_switch.md)
+    * [kubecm clear](/en-us/cli/kubecm_clear.md)
     * [kubecm version](/zh-cn/cli/kubecm_version.md)
 * [贡献](/zh-cn/contribute.md)

--- a/docs/zh-cn/cli/kubecm_clear.md
+++ b/docs/zh-cn/cli/kubecm_clear.md
@@ -1,0 +1,34 @@
+## kubecm clear
+
+清理 kubeconfig 中失效的 `context`, `cluster` 和 `user`
+
+### 简介
+
+默认情况下会清理 `~/.kube/confg` 中失效的 `context`, `cluster` 和 `user`，后跟非默认的 kubeconfig 路径，可以清理指定文件，支持指定多文件。
+
+```
+kubecm clear
+```
+
+### 示例
+
+```
+
+# Clear lapsed context, cluster and user (default is /Users/saybot/.kube/config)
+kubecm clear
+# 自定义清理
+kubecm clear config.yaml test.yaml
+
+```
+
+### Options
+
+```
+  -h, --help   help for clear
+```
+
+### Options inherited from parent commands
+
+```
+      --config string   path of kubeconfig (default "/Users/saybot/.kube/config")
+```

--- a/docs/zh-cn/cli/kubecm_ls.md
+++ b/docs/zh-cn/cli/kubecm_ls.md
@@ -1,10 +1,10 @@
 ## kubecm ls
 
-列举所有 Context
+列举 Context
 
 ### 简介
 
-列举所有 Context
+列举 Context，可以通过后跟关键字，筛选出包含该关键字的 context，支持多关键字筛选，关键字之间是 `或` 的关系。
 
 ```
 kubecm ls
@@ -18,6 +18,8 @@ kubecm ls
 kubecm ls
 # Aliases
 kubecm l
+# Filter out keywords(Multi-keyword support)
+kubecm ls kind k3s
 
 ```
 

--- a/docs/zh-cn/cli/kubecm_switch.md
+++ b/docs/zh-cn/cli/kubecm_switch.md
@@ -7,6 +7,7 @@
 
 交互式切换 Kube Context
 
+**可以通过键入 `\` 来进行搜索**
 
 ```
 kubecm switch [flags]

--- a/go.mod
+++ b/go.mod
@@ -6,6 +6,7 @@ require (
 	github.com/bndr/gotabulate v1.1.2
 	github.com/daviddengcn/go-colortext v1.0.0
 	github.com/imdario/mergo v0.3.7
+	github.com/kylelemons/godebug v1.1.0 // indirect
 	github.com/manifoldco/promptui v0.3.2
 	github.com/rsteube/cobra-zsh-gen v1.1.0
 	github.com/spf13/cobra v1.0.0

